## TOPIC: Classification of Down Syndrome of Mice Protein Dataset
Team Members:
- PRERNA SHARMA-20BCE1407
- TANVI SHRIKHANDE- 20BCE1710
- DEEPAVARSHINI T- 20BCE1661

DATASET SOURCE LINK: 
https://archive.ics.uci.edu/ml/datasets/Mice+Protein+Expression


In [1]:
#importing necessarry libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import time

## Data Background
Expression levels of 77 proteins measured in the cerebral cortex of 8 classes of control and Down syndrome mice exposed to context fear conditioning, a task used to assess associative learning.

In [2]:
df=pd.read_csv('Data_Cortex_Nuclear.csv')
pd.set_option('display.max_row',df.shape[0])
pd.set_option('display.max_column',df.shape[1])
df.head()

MouseID  DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N  \
0   309_1  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830   
1   309_2  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636   
2   309_3  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011   
3   309_4  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886   
4   309_5  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106   

    pBRAF_N  pCAMKII_N   pCREB_N    pELK_N    pERK_N    pJNK_N    PKCA_N  \
0  0.177565   2.373744  0.232224  1.750936  0.687906  0.306382  0.402698   
1  0.172817   2.292150  0.226972  1.596377  0.695006  0.299051  0.385987   
2  0.175722   2.283337  0.230247  1.561316  0.677348  0.291276  0.381002   
3  0.176463   2.152301  0.207004  1.595086  0.583277  0.296729  0.377087   
4  0.173627   2.134014  0.192158  1.504230  0.550960  0.286961  0.363502   

     pMEK_N    pNR1_N   pNR2A_N   pNR2B_N  pPKCAB_N    pRSK_N     AKT_N  \
0  0.296927  1.022060  0.605673  1.877684  2.308745  0.441599  0.859366   
1  0.281319  0.956676  0.587559  1.725774  2.043037  0.445222  0.834659   
2  0.281710  1.003635  0.602449  1.731873  2.017984  0.467668  0.814329   
3  0.313832  0.875390  0.520293  1.566852  2.132754  0.477671  0.727705   
4  0.277964  0.864912  0.507990  1.480059  2.013697  0.483416  0.687794   

     BRAF_N  CAMKII_N    CREB_N     ELK_N     ERK_N   GSK3B_N     JNK_N  \
0  0.416289  0.369608  0.178944  1.866358  3.685247  1.537227  0.264526   
1  0.400364  0.356178  0.173680  1.761047  3.485287  1.509249  0.255727   
2  0.399847  0.368089  0.173905  1.765544  3.571456  1.501244  0.259614   
3  0.385639  0.362970  0.179449  1.286277  2.970137  1.419710  0.259536   
4  0.367531  0.355311  0.174836  1.324695  2.896334  1.359876  0.250705   

      MEK_N    TRKA_N     RSK_N     APP_N  Bcatenin_N    SOD1_N    MTOR_N  \
0  0.319677  0.813866  0.165846  0.453910    3.037621  0.369510  0.458539   
1  0.304419  0.780504  0.157194  0.430940    2.921882  0.342279  0.423560   
2  0.311747  0.785154  0.160895  0.423187    2.944136  0.343696  0.425005   
3  0.279218  0.734492  0.162210  0.410615    2.500204  0.344509  0.429211   
4  0.273667  0.702699  0.154827  0.398550    2.456560  0.329126  0.408755   

      P38_N   pMTOR_N   DSCR1_N   AMPKA_N    NR2B_N   pNUMB_N  RAPTOR_N  \
0  0.335336  0.825192  0.576916  0.448099  0.586271  0.394721  0.339571   
1  0.324835  0.761718  0.545097  0.420876  0.545097  0.368255  0.321959   
2  0.324852  0.757031  0.543620  0.404630  0.552994  0.363880  0.313086   
3  0.330121  0.746980  0.546763  0.386860  0.547849  0.366771  0.328492   
4  0.313415  0.691956  0.536860  0.360816  0.512824  0.351551  0.312206   

    TIAM1_N  pP70S6_N    NUMB_N   P70S6_N  pGSK3B_N   pPKCG_N    CDK5_N  \
0  0.482864  0.294170  0.182150  0.842725  0.192608  1.443091  0.294700   
1  0.454519  0.276431  0.182086  0.847615  0.194815  1.439460  0.294060   
2  0.447197  0.256648  0.184388  0.856166  0.200737  1.524364  0.301881   
3  0.442650  0.398534  0.161768  0.760234  0.184169  1.612382  0.296382   
4  0.419095  0.393447  0.160200  0.768113  0.185718  1.645807  0.296829   

       S6_N  ADARB1_N  AcetylH3K9_N    RRP1_N     BAX_N     ARC_N   ERBB4_N  \
0  0.354605  1.339070      0.170119  0.159102  0.188852  0.106305  0.144989   
1  0.354548  1.306323      0.171427  0.158129  0.184570  0.106592  0.150471   
2  0.386087  1.279600      0.185456  0.148696  0.190532  0.108303  0.145330   
3  0.290680  1.198765      0.159799  0.166112  0.185323  0.103184  0.140656   
4  0.309345  1.206995      0.164650  0.160687  0.188221  0.104784  0.141983   

     nNOS_N     Tau_N    GFAP_N   GluR3_N   GluR4_N    IL1B_N   P3525_N  \
0  0.176668  0.125190  0.115291  0.228043  0.142756  0.430957  0.247538   
1  0.178309  0.134275  0.118235  0.238073  0.142037  0.457156  0.257632   
2  0.176213  0.132560  0.117760  0.244817  0.142445  0.510472  0.255343   
3  0.163804  0.123210  0.117439  0.234947  0.145068  0.430996  0

## Aim:
- Understand the data and perform data preprocessing
- Recognize protiens that influence downs syndrome 
- Implement various classification algorithms on selected protiens to achieve maximum accuracy

## Data Preprocessing

In [3]:
#checking datatypes and presence of null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1080 entries, 0 to 1079
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   MouseID          1080 non-null   object 
 1   DYRK1A_N         1077 non-null   float64
 2   ITSN1_N          1077 non-null   float64
 3   BDNF_N           1077 non-null   float64
 4   NR1_N            1077 non-null   float64
 5   NR2A_N           1077 non-null   float64
 6   pAKT_N           1077 non-null   float64
 7   pBRAF_N          1077 non-null   float64
 8   pCAMKII_N        1077 non-null   float64
 9   pCREB_N          1077 non-null   float64
 10  pELK_N           1077 non-null   float64
 11  pERK_N           1077 non-null   float64
 12  pJNK_N           1077 non-null   float64
 13  PKCA_N           1077 non-null   float64
 14  pMEK_N           1077 non-null   float64
 15  pNR1_N           1077 non-null   float64
 16  pNR2A_N          1077 non-null   float64
 17  pNR2B_N       

We see that the datatypes are suitable and that many columns representing protiens have missing values.

In [4]:
df.dtypes.value_counts()  #number of different datatype variables in dataset

float64    77
object      5
dtype: int64

In [5]:
#dimentionality check
print('There are' , df.shape[0] , 'rows')
print('There are' , df.shape[1] , 'columns')

There are 1080 rows
There are 82 columns


In [6]:
#replacing null values with means of respective columns for further computation
for i in df.columns[df.isnull().any(axis=0)]:     #---Applying Only on variables with NaN values
    df[i].fillna(df[i].mean(),inplace=True)


In [7]:
#checking if any null values are still present
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1080 entries, 0 to 1079
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   MouseID          1080 non-null   object 
 1   DYRK1A_N         1080 non-null   float64
 2   ITSN1_N          1080 non-null   float64
 3   BDNF_N           1080 non-null   float64
 4   NR1_N            1080 non-null   float64
 5   NR2A_N           1080 non-null   float64
 6   pAKT_N           1080 non-null   float64
 7   pBRAF_N          1080 non-null   float64
 8   pCAMKII_N        1080 non-null   float64
 9   pCREB_N          1080 non-null   float64
 10  pELK_N           1080 non-null   float64
 11  pERK_N           1080 non-null   float64
 12  pJNK_N           1080 non-null   float64
 13  PKCA_N           1080 non-null   float64
 14  pMEK_N           1080 non-null   float64
 15  pNR1_N           1080 non-null   float64
 16  pNR2A_N          1080 non-null   float64
 17  pNR2B_N       

**We see all null values have been replaced with column means**

In [8]:
#we only need our feature variables(77 protiens) and our target variable(class)..so we can drop the remaining unwanted columns
df.drop(columns=['MouseID','Genotype', 'Treatment', 'Behavior'], axis=1,inplace=True)

In [9]:
df.columns

Index(['DYRK1A_N', 'ITSN1_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pBRAF_N',
       'pCAMKII_N', 'pCREB_N', 'pELK_N', 'pERK_N', 'pJNK_N', 'PKCA_N',
       'pMEK_N', 'pNR1_N', 'pNR2A_N', 'pNR2B_N', 'pPKCAB_N', 'pRSK_N', 'AKT_N',
       'BRAF_N', 'CAMKII_N', 'CREB_N', 'ELK_N', 'ERK_N', 'GSK3B_N', 'JNK_N',
       'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'SOD1_N', 'MTOR_N',
       'P38_N', 'pMTOR_N', 'DSCR1_N', 'AMPKA_N', 'NR2B_N', 'pNUMB_N',
       'RAPTOR_N', 'TIAM1_N', 'pP70S6_N', 'NUMB_N', 'P70S6_N', 'pGSK3B_N',
       'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N',
       'BAX_N', 'ARC_N', 'ERBB4_N', 'nNOS_N', 'Tau_N', 'GFAP_N', 'GluR3_N',
       'GluR4_N', 'IL1B_N', 'P3525_N', 'pCASP9_N', 'PSD95_N', 'SNCA_N',
       'Ubiquitin_N', 'pGSK3B_Tyr216_N', 'SHH_N', 'BAD_N', 'BCL2_N', 'pS6_N',
       'pCFOS_N', 'SYP_N', 'H3AcK18_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N',
       'class'],
      dtype='object')

1. Normal mouse group
In this data group we want to obtain important proteins
that influence the success / failure of the learning
process in normal mice. The data classes used are cCS-m, c-CS-s, c-SC-m, and c-SC-s. 

In [10]:
# #as all our data is continuous except the class label, using categorical encoder for mice_class
# # Import label encoder
# from sklearn import preprocessing
# df.rename(columns = {'class':'mice_class'}, inplace = True)
# # label_encoder object knows how to understand word labels.
# label_encoder = preprocessing.LabelEncoder()
# # Encode labels in column 'species'.
# df['mice_class']= label_encoder.fit_transform(df['mice_class'])
  
# df['mice_class'].unique()

In [11]:
df.rename(columns = {'class':'mice_class'}, inplace = True)
group1=df.query('mice_class == "c-CS-m" or mice_class=="c-CS-s" or mice_class=="c-SC-m" or mice_class == "c-SC-s"')
group1.head()

DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N   pBRAF_N  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pCAMKII_N   pCREB_N    pELK_N    pERK_N    pJNK_N    PKCA_N    pMEK_N  \
0   2.373744  0.232224  1.750936  0.687906  0.306382  0.402698  0.296927   
1   2.292150  0.226972  1.596377  0.695006  0.299051  0.385987  0.281319   
2   2.283337  0.230247  1.561316  0.677348  0.291276  0.381002  0.281710   
3   2.152301  0.207004  1.595086  0.583277  0.296729  0.377087  0.313832   
4   2.134014  0.192158  1.504230  0.550960  0.286961  0.363502  0.277964   

     pNR1_N   pNR2A_N   pNR2B_N  pPKCAB_N    pRSK_N     AKT_N    BRAF_N  \
0  1.022060  0.605673  1.877684  2.308745  0.441599  0.859366  0.416289   
1  0.956676  0.587559  1.725774  2.043037  0.445222  0.834659  0.400364   
2  1.003635  0.602449  1.731873  2.017984  0.467668  0.814329  0.399847   
3  0.875390  0.520293  1.566852  2.132754  0.477671  0.727705  0.385639   
4  0.864912  0.507990  1.480059  2.013697  0.483416  0.687794  0.367531   

   CAMKII_N    CREB_N     ELK_N     ERK_N   GSK3B_N     JNK_N     MEK_N  \
0  0.369608  0.178944  1.866358  3.685247  1.537227  0.264526  0.319677   
1  0.356178  0.173680  1.761047  3.485287  1.509249  0.255727  0.304419   
2  0.368089  0.173905  1.765544  3.571456  1.501244  0.259614  0.311747   
3  0.362970  0.179449  1.286277  2.970137  1.419710  0.259536  0.279218   
4  0.355311  0.174836  1.324695  2.896334  1.359876  0.250705  0.273667   

     TRKA_N     RSK_N     APP_N  Bcatenin_N    SOD1_N    MTOR_N     P38_N  \
0  0.813866  0.165846  0.453910    3.037621  0.369510  0.458539  0.335336   
1  0.780504  0.157194  0.430940    2.921882  0.342279  0.423560  0.324835   
2  0.785154  0.160895  0.423187    2.944136  0.343696  0.425005  0.324852   
3  0.734492  0.162210  0.410615    2.500204  0.344509  0.429211  0.330121   
4  0.702699  0.154827  0.398550    2.456560  0.329126  0.408755  0.313415   

    pMTOR_N   DSCR1_N   AMPKA_N    NR2B_N   pNUMB_N  RAPTOR_N   TIAM1_N  \
0  0.825192  0.576916  0.448099  0.586271  0.394721  0.339571  0.482864   
1  0.761718  0.545097  0.420876  0.545097  0.368255  0.321959  0.454519   
2  0.757031  0.543620  0.404630  0.552994  0.363880  0.313086  0.447197   
3  0.746980  0.546763  0.386860  0.547849  0.366771  0.328492  0.442650   
4  0.691956  0.536860  0.360816  0.512824  0.351551  0.312206  0.419095   

   pP70S6_N    NUMB_N   P70S6_N  pGSK3B_N   pPKCG_N    CDK5_N      S6_N  \
0  0.294170  0.182150  0.842725  0.192608  1.443091  0.294700  0.354605   
1  0.276431  0.182086  0.847615  0.194815  1.439460  0.294060  0.354548   
2  0.256648  0.184388  0.856166  0.200737  1.524364  0.301881  0.386087   
3  0.398534  0.161768  0.760234  0.184169  1.612382  0.296382  0.290680   
4  0.393447  0.160200  0.768113  0.185718  1.645807  0.296829  0.309345   

   ADARB1_N  AcetylH3K9_N    RRP1_N     BAX_N     ARC_N   ERBB4_N    nNOS_N  \
0  1.339070      0.170119  0.159102  0.188852  0.106305  0.144989  0.176668   
1  1.306323      0.171427  0.158129  0.184570  0.106592  0.150471  0.178309   
2  1.279600      0.185456  0.148696  0.190532  0.108303  0.145330  0.176213   
3  1.198765      0.159799  0.166112  0.185323  0.103184  0.140656  0.163804   
4  1.206995      0.164650  0.160687  0.188221  0.104784  0.141983  0.167710   

      Tau_N    GFAP_N   GluR3_N   GluR4_N    IL1B_N   P3525_N  pCASP9_N  \
0  0.125190  0.115291  0.228043  0.142756  0.430957  0.247538  1.603310   
1  0.134275  0.118235  0.238073  0.142037  0.457156  0.257632  1.671738   
2  0.132560  0.117760  0.244817  0.142445  0.510472  0.255343  1.663550   
3  0.123210  0.117439  0.234947  0.145068  0.430996  

In [12]:
#as all our data is continuous except the class label, using categorical encoder for mice_class
# Import label encoder
from sklearn import preprocessing
# label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
# Encode labels in column 'species'.
group1['mice_class']= label_encoder.fit_transform(group1['mice_class'])
  
group1['mice_class'].unique()

C:\Users\prern\AppData\Local\Temp/ipykernel_12972/3159675617.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  group1['mice_class']= label_encoder.fit_transform(group1['mice_class'])


array([0, 2, 1, 3])

In [13]:
import seaborn as sns
import matplotlib.pyplot as plt
from random import randint
%matplotlib inline 
import warnings
warnings.filterwarnings("ignore")


from sklearn import svm
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold, cross_val_score
def split(df,label):
    X_tr, X_te, Y_tr, Y_te = train_test_split(df, label, test_size=0.25, random_state=42)
    return X_tr, X_te, Y_tr, Y_te


classifiers = ['LinearSVM', 'RadialSVM', 
               'Logistic',  'RandomForest', 
               'AdaBoost',  'DecisionTree', 
               'KNeighbors']

models = [svm.SVC(kernel='linear'),
          svm.SVC(kernel='rbf'),
          LogisticRegression(max_iter = 100),
          RandomForestClassifier(n_estimators=15, random_state=0),
          AdaBoostClassifier(random_state = 0),
          DecisionTreeClassifier(random_state=0),
          KNeighborsClassifier(),
            ]


def acc_score(df,label):
    Score = pd.DataFrame({"Classifier":classifiers})
    j = 0
    acc = []
    X_train,X_test,Y_train,Y_test = split(df,label)
    for i in models:
        model = i
        model.fit(X_train,Y_train)
        predictions = model.predict(X_test)
        acc.append(accuracy_score(Y_test,predictions))
        j = j+1     
    Score["Accuracy"] = acc
    Score.sort_values(by="Accuracy", ascending=False,inplace = True)
    Score.reset_index(drop=True, inplace=True)
    return Score

def plot(score,x,y,c = "b"):
    gen = [1,2,3,4,5]
    plt.figure(figsize=(6,4))
    ax = sns.pointplot(x=gen, y=score,color = c )
    ax.set(xlabel="Generation", ylabel="Accuracy")
    ax.set(ylim=(x,y))

In [14]:
def initilization_of_population(size,n_feat):
    population = []
    for i in range(size):
        chromosome = np.ones(n_feat,dtype=np.bool)     
        chromosome[:int(0.3*n_feat)]=False             
        np.random.shuffle(chromosome)
        population.append(chromosome)
    return population


def fitness_score(population):
    scores = []
    for chromosome in population:
        logmodel.fit(X_train.iloc[:,chromosome],Y_train)         
        predictions = logmodel.predict(X_test.iloc[:,chromosome])
        scores.append(accuracy_score(Y_test,predictions))
    scores, population = np.array(scores), np.array(population) 
    inds = np.argsort(scores)                                    
    return list(scores[inds][::-1]), list(population[inds,:][::-1]) 


def selection(pop_after_fit,n_parents):
    population_nextgen = []
    for i in range(n_parents):
        population_nextgen.append(pop_after_fit[i])
    return population_nextgen


def crossover(pop_after_sel):
    pop_nextgen = pop_after_sel
    for i in range(0,len(pop_after_sel),2):
        new_par = []
        child_1 , child_2 = pop_nextgen[i] , pop_nextgen[i+1]
        new_par = np.concatenate((child_1[:len(child_1)//2],child_2[len(child_1)//2:]))
        pop_nextgen.append(new_par)
    return pop_nextgen


def mutation(pop_after_cross,mutation_rate,n_feat):   
    mutation_range = int(mutation_rate*n_feat)
    pop_next_gen = []
    for n in range(0,len(pop_after_cross)):
        chromo = pop_after_cross[n]
        rand_posi = [] 
        for i in range(0,mutation_range):
            pos = randint(0,n_feat-1)
            rand_posi.append(pos)
        for j in rand_posi:
            chromo[j] = not chromo[j]  
        pop_next_gen.append(chromo)
    return pop_next_gen

def generations(df,label,size,n_feat,n_parents,mutation_rate,n_gen,X_train,
                                   X_test, Y_train, Y_test):
    best_chromo= []
    best_score= []
    population_nextgen=initilization_of_population(size,n_feat)
    for i in range(n_gen):
        scores, pop_after_fit = fitness_score(population_nextgen)
        print('Best score in generation',i+1,':',scores[:1])  #2
        pop_after_sel = selection(pop_after_fit,n_parents)
        pop_after_cross = crossover(pop_after_sel)
        population_nextgen = mutation(pop_after_cross,mutation_rate,n_feat)
        best_chromo.append(pop_after_fit[0])
        best_score.append(scores[0])
    return best_chromo,best_score

In [15]:
#checking accuracy
X=group1.drop(['mice_class'],axis=1)
Y=group1['mice_class']
score1 = acc_score(X,Y)
score1

Classifier  Accuracy
0  RandomForest  0.993007
1    KNeighbors  0.965035
2     LinearSVM  0.937063
3      Logistic  0.895105
4  DecisionTree  0.874126
5     RadialSVM  0.734266
6      AdaBoost  0.524476

#### Choosing the best classifier for further calculations

In [16]:
logmodel = RandomForestClassifier(n_estimators=20, random_state=0)
X_train,X_test, Y_train, Y_test = split(X,Y)
chromo_df,score_df=generations(X,Y,size=80,n_feat=X.shape[1],n_parents=64,mutation_rate=0.20,n_gen=5,
                         X_train = X_train,X_test = X_test,Y_train = Y_train,Y_test = Y_test)

Best score in generation 1 : [1.0]
Best score in generation 2 : [1.0]
Best score in generation 3 : [1.0]
Best score in generation 4 : [1.0]
Best score in generation 5 : [1.0]


In [17]:
imp_features_group1=[]
for i in range(len(X.columns)):
    if(chromo_df[-1][i]):
        imp_features_group1.append(X.columns[i])
print(imp_features_group1)
        

['DYRK1A_N', 'BDNF_N', 'pAKT_N', 'pCAMKII_N', 'pCREB_N', 'pERK_N', 'PKCA_N', 'pMEK_N', 'pNR2A_N', 'pNR2B_N', 'pPKCAB_N', 'pRSK_N', 'CREB_N', 'ELK_N', 'ERK_N', 'MEK_N', 'TRKA_N', 'APP_N', 'SOD1_N', 'AMPKA_N', 'pNUMB_N', 'RAPTOR_N', 'pP70S6_N', 'NUMB_N', 'pGSK3B_N', 'pPKCG_N', 'S6_N', 'AcetylH3K9_N', 'BAX_N', 'Tau_N', 'GFAP_N', 'GluR3_N', 'GluR4_N', 'IL1B_N', 'P3525_N', 'pCASP9_N', 'PSD95_N', 'SNCA_N', 'SHH_N', 'BCL2_N', 'pS6_N', 'pCFOS_N', 'H3AcK18_N', 'EGR1_N', 'H3MeK4_N']


In [18]:
len(imp_features_group1)

45

In [19]:
from genetic_selection import GeneticSelectionCV
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
estimator = DecisionTreeClassifier()
model = GeneticSelectionCV(
    estimator, cv=5, verbose=0,
    scoring="accuracy", max_features=50,
    n_population=100, crossover_proba=0.5,
    mutation_proba=0.2, n_generations=50,
    crossover_independent_proba=0.5,
    mutation_independent_proba=0.04,
    tournament_size=3, n_gen_no_change=10,
    caching=True, n_jobs=-1)
model = model.fit(X, Y)
print('Features:', X.columns[model.support_])

Features: Index(['ITSN1_N', 'NR2A_N', 'pCAMKII_N', 'pJNK_N', 'pNR1_N', 'pNR2A_N',
       'pPKCAB_N', 'GSK3B_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N',
       'SOD1_N', 'MTOR_N', 'P38_N', 'TIAM1_N', 'pPKCG_N', 'ARC_N', 'ERBB4_N',
       'nNOS_N', 'GFAP_N', 'IL1B_N', 'P3525_N', 'BAD_N', 'H3AcK18_N',
       'CaNA_N'],
      dtype='object')


2. Trisomy mouse group (success vs. failure)
In this data group we want to obtain important proteins
that influence the success of the trisomy rat learning
process given memantine treatment. The data classes
used are t-CS-m and t-CS-s.

In [22]:
group2=df.query('mice_class == "t-CS-m" or mice_class== "t-CS-s"')
group2

DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N   pBRAF_N  \
570  0.838053  1.006324  0.364953  2.603061  4.207680  0.202548  0.164055   
571  0.938245  1.077099  0.366149  2.703612  4.484228  0.197465  0.169923   
572  0.956488  1.108441  0.384049  2.662564  4.606357  0.203799  0.171625   
573  0.731738  0.944241  0.324365  2.320243  3.621250  0.198649  0.162240   
574  0.815503  1.005562  0.335976  2.379290  3.641657  0.217870  0.164142   
575  0.807475  0.925253  0.334827  2.327989  4.018038  0.206437  0.165173   
576  0.668978  0.903673  0.312435  2.291341  3.594778  0.207110  0.169937   
577  0.776663  0.916962  0.327057  2.407933  3.627103  0.209446  0.173021   
578  0.646086  0.942658  0.313107  2.259709  3.694630  0.217536  0.170813   
579  0.602820  0.835561  0.301044  2.133126  3.064457  0.215345  0.184765   
580  0.625976  0.806894  0.327938  2.217863  3.521615  0.236526  0.190059   
581  0.605380  0.875750  0.312754  2.184633  3.676795  0.225276  0.189085   
582  0.555122  0.726229  0.278319  2.097249  2.897553  0.222222  0.174356   
583  0.543226  0.757768  0.294018  2.014162  3.134432  0.211795  0.170788   
584  0.574555  0.714868  0.278910  2.026818  3.281270  0.200815  0.166059   
585  0.641872  0.859113  0.327980  2.652611  4.201576  0.190246  0.159704   
586  0.673085  0.877280  0.341103  2.696291  4.180077  0.194568  0.161127   
587  0.660194  0.828580  0.341930  2.662217  4.581311  0.198625  0.159486   
588  0.571500  0.747993  0.311465  2.450201  3.827270  0.200075  0.165454   
589  0.540852  0.738703  0.318659  2.479218  4.044931  0.203418  0.168069   
590  0.649590  0.728996  0.316855  2.486296  3.866675  0.200948  0.186219   
591  0.474654  0.671429  0.294931  2.251920  3.322581  0.204455  0.185868   
592  0.464443  0.668992  0.290523  2.188476  3.500682  0.203184  0.163002   
593  0.542599  0.673647  0.307473  2.346502  3.612724  0.211963  0.171664   
594  0.418401  0.535349  0.279319  1.909812  2.686653  0.190453  0.159564   
595  0.406920  0.520235  0.249167  1.799556  2.358673  0.176956  0.152198   
596  0.445735  0.584778  0.268377  1.993759  2.746879  0.199202  0.160888   
597  0.511218  0.657280  0.322344  2.196658  3.155678  0.243590  0.194597   
598  0.392627  0.503042  0.251432  1.726020  2.377595  0.187903  0.149606   
599  0.440879  0.539513  0.263418  1.896811  2.458903  0.208358  0.163003   
600  0.449650  0.648032  0.261531  1.789181  2.991789  0.181961  0.143444   
601  0.486867  0.681813  0.250591  1.835927  3.239014  0.181750  0.147516   
602  0.524219  0.700130  0.257682  1.954036  3.281380  0.194922  0.154818   
603  0.427303  0.579741  0.233023  1.637081  2.612426  0.184982  0.143280   
604  0.434018  0.563343  0.241056  1.720235  2.727859  0.187097  0.146628   
605  0.467166  0.628451  0.255966  1.785681  2.848698  0.198253  0.158634   
606  0.369484  0.485913  0.228785  1.441955  2.103700  0.184318  0.149355   
607  0.403018  0.530091  0.237377  1.603317  2.509409  0.197503  0.154090   
608  0.408687  0.569884  0.243436  1.633784  2.580888  0.209653  0.161969   
609  0.413552  0.502295  0.231257  1.499454  2.086339  0.208743  0.166995   
610  0.418160  0.523821  0.251415  1.564858  2.319575  0.224764  0.188443   
611  0.407619  0.540952  0.255000  1.603571  2.356667  0.246190  0.177619   
612  0.358105  0.463231  0.293371  1.465821  2.039876  0.210254  0.173226   
613  0.377395  0.486590  0.253147  1.554735  2.141215  0.223864  0.195676   
614  0.495831  0.500000  0.290439  1.569483  2.217343  0.229294  0.189550   
615  0.853106  1.071848  0.280459  2.080645  3.850307  0.181285  0.154625   
616  0.872788  0.980173  0.274917  2.052562  3.808916  0.183500  0.149701   
617  0.888564  1.071410  0.293750  2.045479  3.782979  0.175133  0.150133   
618  0.765430  0.926483  0.264490  1.943580  3.380407  0.186015  0.159173   
619  0.799383  0.889155  0.259551  1.938667  3.418023  0.188282  0.152304   
620  0.712609  0.908283  0.253353  1.922368  3.242622  0.184943  0.148726   
6

In [23]:
#as all our data is continuous except the class label, using categorical encoder for mice_class
# Import label encoder
from sklearn import preprocessing
# label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
# Encode labels in column 'species'.
group2['mice_class']= label_encoder.fit_transform(group2['mice_class'])
  
group2['mice_class'].unique()

array([0, 1])

In [24]:
#checking accuracy
X1=group2.drop(['mice_class'],axis=1)
Y1=group2['mice_class']
score2 = acc_score(X1,Y1)
score2

Classifier  Accuracy
0     LinearSVM  0.983333
1  RandomForest  0.983333
2      AdaBoost  0.983333
3      Logistic  0.950000
4  DecisionTree  0.933333
5    KNeighbors  0.916667
6     RadialSVM  0.800000

#### Choosing the best classifier for further calculations

In [25]:
#implementing best model
from sklearn import svm #Import svm model
logmodel = svm.SVC(kernel='linear') # rbf Kernel
#Train the model using the training sets
X_train,X_test, Y_train, Y_test = split(X1,Y1)
chromo_df,score_df=generations(X,Y,size=80,n_feat=X1.shape[1],n_parents=64,mutation_rate=0.20,n_gen=5,
                         X_train = X_train,X_test = X_test,Y_train = Y_train,Y_test = Y_test)

Best score in generation 1 : [1.0]
Best score in generation 2 : [1.0]
Best score in generation 3 : [1.0]
Best score in generation 4 : [1.0]
Best score in generation 5 : [1.0]


In [26]:
imp_features_group2=[]
for i in range(len(X1.columns)):
    if(chromo_df[-1][i]):
        imp_features_group2.append(X1.columns[i])
print(imp_features_group2)
        

['BDNF_N', 'NR2A_N', 'pAKT_N', 'pBRAF_N', 'pCAMKII_N', 'pERK_N', 'PKCA_N', 'pMEK_N', 'pNR1_N', 'pNR2B_N', 'pPKCAB_N', 'pRSK_N', 'BRAF_N', 'CREB_N', 'ERK_N', 'JNK_N', 'MEK_N', 'TRKA_N', 'RSK_N', 'Bcatenin_N', 'P38_N', 'DSCR1_N', 'AMPKA_N', 'pNUMB_N', 'TIAM1_N', 'pP70S6_N', 'P70S6_N', 'pGSK3B_N', 'pPKCG_N', 'CDK5_N', 'S6_N', 'AcetylH3K9_N', 'RRP1_N', 'ARC_N', 'nNOS_N', 'GFAP_N', 'IL1B_N', 'PSD95_N', 'Ubiquitin_N', 'pGSK3B_Tyr216_N', 'SHH_N', 'BAD_N', 'pCFOS_N', 'H3AcK18_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N']


In [27]:
len(imp_features_group2)

47

In [28]:
from genetic_selection import GeneticSelectionCV
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
estimator = DecisionTreeClassifier()
model = GeneticSelectionCV(
    estimator, cv=5, verbose=0,
    scoring="accuracy", max_features=50,
    n_population=100, crossover_proba=0.5,
    mutation_proba=0.2, n_generations=50,
    crossover_independent_proba=0.5,
    mutation_independent_proba=0.04,
    tournament_size=3, n_gen_no_change=10,
    caching=True, n_jobs=-1)
model = model.fit(X1, Y1)
print('Features:', X1.columns[model.support_])

Features: Index(['NR2A_N', 'pJNK_N', 'pPKCAB_N', 'BRAF_N', 'CAMKII_N', 'ERK_N', 'MEK_N',
       'SOD1_N', 'MTOR_N', 'AMPKA_N', 'NR2B_N', 'P70S6_N', 'S6_N', 'ARC_N',
       'nNOS_N', 'pS6_N', 'CaNA_N'],
      dtype='object')


3. Normal mouse group with trisomy mice
In this data group we want to obtain important proteins
that influence the failure of trisomy mice in studying
their environment and compared to normal mice who
have successfully studied their environment. The data
classes used are t-CS-s, c-CS-m, and c-CS-s. 

In [29]:
group3=df.query('mice_class == "t-CS-s" or mice_class== "c-CS-m" or mice_class=="c-CS-s"')
group3

DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N   pBRAF_N  \
0    0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1    0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2    0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3    0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4    0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   
5    0.447506  0.628176  0.367388  2.385939  4.807635  0.218578  0.176233   
6    0.428033  0.573696  0.342709  2.334224  4.473130  0.225173  0.184004   
7    0.416923  0.564036  0.327703  2.260135  4.268735  0.214834  0.179668   
8    0.386311  0.538428  0.317720  2.125725  4.063950  0.207222  0.167778   
9    0.380827  0.499294  0.362462  2.096266  3.598587  0.227649  0.188093   
10   0.366511  0.513278  0.327792  2.072572  3.661058  0.229475  0.193797   
11   0.364154  0.499411  0.355124  2.006871  3.466627  0.216333  0.204358   
12   0.364873  0.482102  0.312539  1.946201  3.349592  0.230898  0.188821   
13   0.381911  0.485914  0.310951  1.958907  3.349290  0.225800  0.188519   
14   0.374410  0.462313  0.344629  1.861162  3.287123  0.221811  0.185459   
15   0.743118  0.862653  0.377742  2.735757  6.067570  0.219049  0.185338   
16   0.711480  0.807054  0.351591  2.546888  5.595574  0.199170  0.165975   
17   0.704633  0.802537  0.350110  2.467733  5.548400  0.205323  0.165058   
18   0.677359  0.770235  0.356397  2.563223  4.975196  0.228087  0.186498   
19   0.591572  0.678768  0.312480  2.164182  4.313938  0.195786  0.161102   
20   0.618517  0.716672  0.319701  2.285938  4.571180  0.206927  0.171598   
21   0.702608  0.699581  0.387750  2.437587  4.449231  0.258500  0.197019   
22   0.598869  0.690271  0.349774  2.308371  4.229412  0.221493  0.187104   
23   0.561867  0.641892  0.308488  2.157728  4.020693  0.216427  0.173775   
24   0.550970  0.561356  0.320853  2.197868  3.558896  0.237770  0.194589   
25   0.538413  0.701865  0.384445  2.482137  4.109706  0.288018  0.223206   
26   0.521127  0.583045  0.304279  2.053415  3.381345  0.251661  0.189476   
27   0.488372  0.540086  0.291310  1.913403  2.873623  0.215422  0.210526   
28   0.514855  0.564791  0.316056  1.957016  2.978824  0.250632  0.175095   
29   0.485100  0.556375  0.287250  1.892166  2.847005  0.223041  0.177573   
30   0.627582  0.954187  0.446958  2.930717  5.915039  0.197897  0.185599   
31   0.651253  0.962021  0.464964  2.992689  5.975028  0.202810  0.188473   
32   0.644346  0.967179  0.470056  3.073847  5.927379  0.205131  0.182590   
33   0.568229  0.812018  0.393465  2.606780  5.808102  0.218211  0.179905   
34   0.587038  0.863728  0.411203  2.757975  6.006657  0.223688  0.189400   
35   0.595904  0.861763  0.421204  2.831169  6.368257  0.233017  0.187687   
36   0.522918  0.721090  0.363541  2.456261  5.191732  0.229179  0.186639   
37   0.530855  0.748072  0.385529  2.573126  5.495063  0.237427  0.187596   
38   0.534999  0.773863  0.391588  2.667701  5.987583  0.246624  0.184541   
39   0.519443  0.736371  0.370568  2.503050  4.947198  0.266489  0.205299   
40   0.508590  0.735699  0.384369  2.545403  5.101001  0.257127  0.201057   
41   0.516538  0.737479  0.383670  2.624646  5.601210  0.259119  0.183897   
42   0.454421  0.685278  0.360301  2.392890  4.500000  0.278715  0.208751   
43   0.471746  0.697323  0.371540  2.472661  4.652025  0.278197  0.215511   
44   0.481273  0.684116  0.360785  2.538809  5.008123  0.287229  0.224504   
45   0.560614  0.763028  0.402962  2.933516  6.317279  0.192320  0.167636   
46   0.516108  0.749622  0.401297  2.919243  6.306811  0.192541  0.162270   
47   0.507321  0.770026  0.417959  2.925818  6.766150  0.201012  0.177756   
48   0.505093  0.695549  0.376029  2.915585  5.917957  0.226734  0.174271   
49   0.479120  0.689170  0.381125  2.913558  6.066537  0.221882  0.167177   
50   0.457992  0.703492  0.385843  2.904392  6.318467  0.216858  0.160666   
5

In [30]:
#as all our data is continuous except the class label, using categorical encoder for mice_class
# Import label encoder
from sklearn import preprocessing
# label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
# Encode labels in column 'species'.
group3['mice_class']= label_encoder.fit_transform(group3['mice_class'])
  
group3['mice_class'].unique()

array([0, 1, 2])

In [31]:
#checking accuracy
X2=group3.drop(['mice_class'],axis=1)
Y2=group3['mice_class']
score3 = acc_score(X2,Y2)
score3

Classifier  Accuracy
0  RandomForest  0.938776
1     LinearSVM  0.918367
2    KNeighbors  0.918367
3      Logistic  0.897959
4      AdaBoost  0.857143
5  DecisionTree  0.806122
6     RadialSVM  0.561224

#### Choosing the best classifier for further calculations

In [32]:
#implementing best model
logmodel = RandomForestClassifier(n_estimators=20, random_state=0)
X_train,X_test, Y_train, Y_test = split(X2,Y2)
chromo_df,score_df=generations(X2,Y2,size=80,n_feat=X2.shape[1],n_parents=64,mutation_rate=0.20,n_gen=5,
                         X_train = X_train,X_test = X_test,Y_train = Y_train,Y_test = Y_test)

Best score in generation 1 : [0.9897959183673469]
Best score in generation 2 : [1.0]
Best score in generation 3 : [0.9897959183673469]
Best score in generation 4 : [1.0]
Best score in generation 5 : [0.9897959183673469]


In [33]:
imp_features_group3=[]
for i in range(len(X2.columns)):
    if(chromo_df[-1][i]):
        imp_features_group3.append(X2.columns[i])
print(imp_features_group3)
        

['ITSN1_N', 'NR2A_N', 'pCREB_N', 'pERK_N', 'PKCA_N', 'pMEK_N', 'pNR1_N', 'pNR2A_N', 'pNR2B_N', 'pRSK_N', 'CAMKII_N', 'CREB_N', 'ERK_N', 'TRKA_N', 'APP_N', 'SOD1_N', 'DSCR1_N', 'AMPKA_N', 'pP70S6_N', 'CDK5_N', 'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N', 'ARC_N', 'GFAP_N', 'IL1B_N', 'P3525_N', 'PSD95_N', 'SNCA_N', 'pGSK3B_Tyr216_N', 'BCL2_N', 'pS6_N', 'pCFOS_N', 'H3AcK18_N', 'H3MeK4_N', 'CaNA_N']


In [34]:
len(imp_features_group3)

36

In [35]:
from genetic_selection import GeneticSelectionCV
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
estimator = DecisionTreeClassifier()
model = GeneticSelectionCV(
    estimator, cv=5, verbose=0,
    scoring="accuracy", max_features=50,
    n_population=100, crossover_proba=0.5,
    mutation_proba=0.2, n_generations=50,
    crossover_independent_proba=0.5,
    mutation_independent_proba=0.04,
    tournament_size=3, n_gen_no_change=10,
    caching=True, n_jobs=-1)
model = model.fit(X2, Y2)
print('Features:', X1.columns[model.support_])

Features: Index(['pBRAF_N', 'pERK_N', 'BRAF_N', 'CAMKII_N', 'ELK_N', 'ERK_N', 'GSK3B_N',
       'MEK_N', 'TRKA_N', 'APP_N', 'pMTOR_N', 'AMPKA_N', 'NR2B_N', 'pNUMB_N',
       'RAPTOR_N', 'pGSK3B_N', 'ARC_N', 'nNOS_N', 'GluR3_N', 'BAD_N', 'SYP_N',
       'EGR1_N', 'H3MeK4_N'],
      dtype='object')


In [36]:
imp_features_df=[]
for i in imp_features_group1:
    if(i in imp_features_group2):
        if(i in imp_features_group3):
            imp_features_df.append(i)
imp_features_df

['pERK_N',
 'PKCA_N',
 'pMEK_N',
 'pNR2B_N',
 'pRSK_N',
 'CREB_N',
 'ERK_N',
 'TRKA_N',
 'AMPKA_N',
 'pP70S6_N',
 'AcetylH3K9_N',
 'GFAP_N',
 'IL1B_N',
 'PSD95_N',
 'pCFOS_N',
 'H3AcK18_N',
 'H3MeK4_N']

In [37]:
df.head()

DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N   pBRAF_N  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pCAMKII_N   pCREB_N    pELK_N    pERK_N    pJNK_N    PKCA_N    pMEK_N  \
0   2.373744  0.232224  1.750936  0.687906  0.306382  0.402698  0.296927   
1   2.292150  0.226972  1.596377  0.695006  0.299051  0.385987  0.281319   
2   2.283337  0.230247  1.561316  0.677348  0.291276  0.381002  0.281710   
3   2.152301  0.207004  1.595086  0.583277  0.296729  0.377087  0.313832   
4   2.134014  0.192158  1.504230  0.550960  0.286961  0.363502  0.277964   

     pNR1_N   pNR2A_N   pNR2B_N  pPKCAB_N    pRSK_N     AKT_N    BRAF_N  \
0  1.022060  0.605673  1.877684  2.308745  0.441599  0.859366  0.416289   
1  0.956676  0.587559  1.725774  2.043037  0.445222  0.834659  0.400364   
2  1.003635  0.602449  1.731873  2.017984  0.467668  0.814329  0.399847   
3  0.875390  0.520293  1.566852  2.132754  0.477671  0.727705  0.385639   
4  0.864912  0.507990  1.480059  2.013697  0.483416  0.687794  0.367531   

   CAMKII_N    CREB_N     ELK_N     ERK_N   GSK3B_N     JNK_N     MEK_N  \
0  0.369608  0.178944  1.866358  3.685247  1.537227  0.264526  0.319677   
1  0.356178  0.173680  1.761047  3.485287  1.509249  0.255727  0.304419   
2  0.368089  0.173905  1.765544  3.571456  1.501244  0.259614  0.311747   
3  0.362970  0.179449  1.286277  2.970137  1.419710  0.259536  0.279218   
4  0.355311  0.174836  1.324695  2.896334  1.359876  0.250705  0.273667   

     TRKA_N     RSK_N     APP_N  Bcatenin_N    SOD1_N    MTOR_N     P38_N  \
0  0.813866  0.165846  0.453910    3.037621  0.369510  0.458539  0.335336   
1  0.780504  0.157194  0.430940    2.921882  0.342279  0.423560  0.324835   
2  0.785154  0.160895  0.423187    2.944136  0.343696  0.425005  0.324852   
3  0.734492  0.162210  0.410615    2.500204  0.344509  0.429211  0.330121   
4  0.702699  0.154827  0.398550    2.456560  0.329126  0.408755  0.313415   

    pMTOR_N   DSCR1_N   AMPKA_N    NR2B_N   pNUMB_N  RAPTOR_N   TIAM1_N  \
0  0.825192  0.576916  0.448099  0.586271  0.394721  0.339571  0.482864   
1  0.761718  0.545097  0.420876  0.545097  0.368255  0.321959  0.454519   
2  0.757031  0.543620  0.404630  0.552994  0.363880  0.313086  0.447197   
3  0.746980  0.546763  0.386860  0.547849  0.366771  0.328492  0.442650   
4  0.691956  0.536860  0.360816  0.512824  0.351551  0.312206  0.419095   

   pP70S6_N    NUMB_N   P70S6_N  pGSK3B_N   pPKCG_N    CDK5_N      S6_N  \
0  0.294170  0.182150  0.842725  0.192608  1.443091  0.294700  0.354605   
1  0.276431  0.182086  0.847615  0.194815  1.439460  0.294060  0.354548   
2  0.256648  0.184388  0.856166  0.200737  1.524364  0.301881  0.386087   
3  0.398534  0.161768  0.760234  0.184169  1.612382  0.296382  0.290680   
4  0.393447  0.160200  0.768113  0.185718  1.645807  0.296829  0.309345   

   ADARB1_N  AcetylH3K9_N    RRP1_N     BAX_N     ARC_N   ERBB4_N    nNOS_N  \
0  1.339070      0.170119  0.159102  0.188852  0.106305  0.144989  0.176668   
1  1.306323      0.171427  0.158129  0.184570  0.106592  0.150471  0.178309   
2  1.279600      0.185456  0.148696  0.190532  0.108303  0.145330  0.176213   
3  1.198765      0.159799  0.166112  0.185323  0.103184  0.140656  0.163804   
4  1.206995      0.164650  0.160687  0.188221  0.104784  0.141983  0.167710   

      Tau_N    GFAP_N   GluR3_N   GluR4_N    IL1B_N   P3525_N  pCASP9_N  \
0  0.125190  0.115291  0.228043  0.142756  0.430957  0.247538  1.603310   
1  0.134275  0.118235  0.238073  0.142037  0.457156  0.257632  1.671738   
2  0.132560  0.117760  0.244817  0.142445  0.510472  0.255343  1.663550   
3  0.123210  0.117439  0.234947  0.145068  0.430996  

In [38]:
#as all our data is continuous except the class label, using categorical encoder for mice_class
# Import label encoder
from sklearn import preprocessing
# label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
# Encode labels in column 'species'.
df['mice_class']= label_encoder.fit_transform(df['mice_class'])
  
df['mice_class'].unique()

array([0, 2, 1, 3, 4, 6, 5, 7])

In [39]:
XX=df[['DYRK1A_N',
 'BDNF_N',
 'NR2A_N',
 'pCAMKII_N',
 'pMEK_N',
 'pPKCAB_N',
 'S6_N',
 'nNOS_N',
 'SNCA_N',
 'H3AcK18_N']]
YY=df['mice_class']

In [40]:
#standardizing the data to improve performance of data models on it
ss = StandardScaler()
XX= ss.fit_transform(XX)
XX

array([[ 0.31271112,  2.2536689 ,  2.30436515, ..., -0.18598559,
        -2.13707513, -1.01161458],
       [ 0.35679793,  1.8802795 ,  1.97676855, ..., -0.12008826,
        -2.07434706, -1.06345775],
       [ 0.33496588,  2.01292763,  1.9091485 , ..., -0.20424585,
        -2.13865527, -1.06512993],
       ...,
       [-0.79192771, -1.72382963, -1.74268127, ...,  0.15382399,
         0.82318544,  2.79900403],
       [-0.82188815, -1.52387571, -1.56741722, ...,  0.49998377,
         1.04463202,  3.06895768],
       [-0.49491588, -1.26845332, -1.34166363, ...,  0.40190703,
         1.02305418,  3.05282386]])

In [41]:
#Splitting the dataset into  training and validation sets
X_train, X_test, Y_train, Y_test = train_test_split(XX, YY, test_size=0.2,
                                                    random_state=1,stratify=YY) #using stratify to preserve the percentage of data in each class

In [42]:
#single deep layer with relu and adam solver
#Initializing the MLPClassifier
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(200,200), max_iter=1000,activation = 'relu',solver='adam',random_state=1)
classifier.fit(X_train, Y_train)
y_pred = classifier.predict(X_test)
#Calculating the accuracy of predictions
def accuracy(confusion_matrix):
    diagonal_sum = confusion_matrix.trace()
    sum_of_all_elements = confusion_matrix.sum()
    return diagonal_sum / sum_of_all_elements
#Importing Confusion Matrix
from sklearn.metrics import confusion_matrix
#Comparing the predictions against the actual observations in y_val
cm = confusion_matrix(y_pred, Y_test)

#Printing the accuracy
print("Accuracy of MLPClassifier : ", accuracy(cm))

Accuracy of MLPClassifier :  0.9629629629629629


In [43]:
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(n_estimators=150, random_state=0)
random_forest.fit(X_train, Y_train)
Y_pred = random_forest.predict(X_test)
accuracy_score(Y_test,Y_pred)

0.9675925925925926

In [44]:
#applying genetic algorithm on whole data
df.head()

DYRK1A_N   ITSN1_N    BDNF_N     NR1_N    NR2A_N    pAKT_N   pBRAF_N  \
0  0.503644  0.747193  0.430175  2.816329  5.990152  0.218830  0.177565   
1  0.514617  0.689064  0.411770  2.789514  5.685038  0.211636  0.172817   
2  0.509183  0.730247  0.418309  2.687201  5.622059  0.209011  0.175722   
3  0.442107  0.617076  0.358626  2.466947  4.979503  0.222886  0.176463   
4  0.434940  0.617430  0.358802  2.365785  4.718679  0.213106  0.173627   

   pCAMKII_N   pCREB_N    pELK_N    pERK_N    pJNK_N    PKCA_N    pMEK_N  \
0   2.373744  0.232224  1.750936  0.687906  0.306382  0.402698  0.296927   
1   2.292150  0.226972  1.596377  0.695006  0.299051  0.385987  0.281319   
2   2.283337  0.230247  1.561316  0.677348  0.291276  0.381002  0.281710   
3   2.152301  0.207004  1.595086  0.583277  0.296729  0.377087  0.313832   
4   2.134014  0.192158  1.504230  0.550960  0.286961  0.363502  0.277964   

     pNR1_N   pNR2A_N   pNR2B_N  pPKCAB_N    pRSK_N     AKT_N    BRAF_N  \
0  1.022060  0.605673  1.877684  2.308745  0.441599  0.859366  0.416289   
1  0.956676  0.587559  1.725774  2.043037  0.445222  0.834659  0.400364   
2  1.003635  0.602449  1.731873  2.017984  0.467668  0.814329  0.399847   
3  0.875390  0.520293  1.566852  2.132754  0.477671  0.727705  0.385639   
4  0.864912  0.507990  1.480059  2.013697  0.483416  0.687794  0.367531   

   CAMKII_N    CREB_N     ELK_N     ERK_N   GSK3B_N     JNK_N     MEK_N  \
0  0.369608  0.178944  1.866358  3.685247  1.537227  0.264526  0.319677   
1  0.356178  0.173680  1.761047  3.485287  1.509249  0.255727  0.304419   
2  0.368089  0.173905  1.765544  3.571456  1.501244  0.259614  0.311747   
3  0.362970  0.179449  1.286277  2.970137  1.419710  0.259536  0.279218   
4  0.355311  0.174836  1.324695  2.896334  1.359876  0.250705  0.273667   

     TRKA_N     RSK_N     APP_N  Bcatenin_N    SOD1_N    MTOR_N     P38_N  \
0  0.813866  0.165846  0.453910    3.037621  0.369510  0.458539  0.335336   
1  0.780504  0.157194  0.430940    2.921882  0.342279  0.423560  0.324835   
2  0.785154  0.160895  0.423187    2.944136  0.343696  0.425005  0.324852   
3  0.734492  0.162210  0.410615    2.500204  0.344509  0.429211  0.330121   
4  0.702699  0.154827  0.398550    2.456560  0.329126  0.408755  0.313415   

    pMTOR_N   DSCR1_N   AMPKA_N    NR2B_N   pNUMB_N  RAPTOR_N   TIAM1_N  \
0  0.825192  0.576916  0.448099  0.586271  0.394721  0.339571  0.482864   
1  0.761718  0.545097  0.420876  0.545097  0.368255  0.321959  0.454519   
2  0.757031  0.543620  0.404630  0.552994  0.363880  0.313086  0.447197   
3  0.746980  0.546763  0.386860  0.547849  0.366771  0.328492  0.442650   
4  0.691956  0.536860  0.360816  0.512824  0.351551  0.312206  0.419095   

   pP70S6_N    NUMB_N   P70S6_N  pGSK3B_N   pPKCG_N    CDK5_N      S6_N  \
0  0.294170  0.182150  0.842725  0.192608  1.443091  0.294700  0.354605   
1  0.276431  0.182086  0.847615  0.194815  1.439460  0.294060  0.354548   
2  0.256648  0.184388  0.856166  0.200737  1.524364  0.301881  0.386087   
3  0.398534  0.161768  0.760234  0.184169  1.612382  0.296382  0.290680   
4  0.393447  0.160200  0.768113  0.185718  1.645807  0.296829  0.309345   

   ADARB1_N  AcetylH3K9_N    RRP1_N     BAX_N     ARC_N   ERBB4_N    nNOS_N  \
0  1.339070      0.170119  0.159102  0.188852  0.106305  0.144989  0.176668   
1  1.306323      0.171427  0.158129  0.184570  0.106592  0.150471  0.178309   
2  1.279600      0.185456  0.148696  0.190532  0.108303  0.145330  0.176213   
3  1.198765      0.159799  0.166112  0.185323  0.103184  0.140656  0.163804   
4  1.206995      0.164650  0.160687  0.188221  0.104784  0.141983  0.167710   

      Tau_N    GFAP_N   GluR3_N   GluR4_N    IL1B_N   P3525_N  pCASP9_N  \
0  0.125190  0.115291  0.228043  0.142756  0.430957  0.247538  1.603310   
1  0.134275  0.118235  0.238073  0.142037  0.457156  0.257632  1.671738   
2  0.132560  0.117760  0.244817  0.142445  0.510472  0.255343  1.663550   
3  0.123210  0.117439  0.234947  0.145068  0.430996  

In [45]:
df.columns

Index(['DYRK1A_N', 'ITSN1_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pBRAF_N',
       'pCAMKII_N', 'pCREB_N', 'pELK_N', 'pERK_N', 'pJNK_N', 'PKCA_N',
       'pMEK_N', 'pNR1_N', 'pNR2A_N', 'pNR2B_N', 'pPKCAB_N', 'pRSK_N', 'AKT_N',
       'BRAF_N', 'CAMKII_N', 'CREB_N', 'ELK_N', 'ERK_N', 'GSK3B_N', 'JNK_N',
       'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'SOD1_N', 'MTOR_N',
       'P38_N', 'pMTOR_N', 'DSCR1_N', 'AMPKA_N', 'NR2B_N', 'pNUMB_N',
       'RAPTOR_N', 'TIAM1_N', 'pP70S6_N', 'NUMB_N', 'P70S6_N', 'pGSK3B_N',
       'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N',
       'BAX_N', 'ARC_N', 'ERBB4_N', 'nNOS_N', 'Tau_N', 'GFAP_N', 'GluR3_N',
       'GluR4_N', 'IL1B_N', 'P3525_N', 'pCASP9_N', 'PSD95_N', 'SNCA_N',
       'Ubiquitin_N', 'pGSK3B_Tyr216_N', 'SHH_N', 'BAD_N', 'BCL2_N', 'pS6_N',
       'pCFOS_N', 'SYP_N', 'H3AcK18_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N',
       'mice_class'],
      dtype='object')

In [46]:
X_df=df[['DYRK1A_N', 'ITSN1_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pBRAF_N',
       'pCAMKII_N', 'pCREB_N', 'pELK_N', 'pERK_N', 'pJNK_N', 'PKCA_N',
       'pMEK_N', 'pNR1_N', 'pNR2A_N', 'pNR2B_N', 'pPKCAB_N', 'pRSK_N', 'AKT_N',
       'BRAF_N', 'CAMKII_N', 'CREB_N', 'ELK_N', 'ERK_N', 'GSK3B_N', 'JNK_N',
       'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'SOD1_N', 'MTOR_N',
       'P38_N', 'pMTOR_N', 'DSCR1_N', 'AMPKA_N', 'NR2B_N', 'pNUMB_N',
       'RAPTOR_N', 'TIAM1_N', 'pP70S6_N', 'NUMB_N', 'P70S6_N', 'pGSK3B_N',
       'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N',
       'BAX_N', 'ARC_N', 'ERBB4_N', 'nNOS_N', 'Tau_N', 'GFAP_N', 'GluR3_N',
       'GluR4_N', 'IL1B_N', 'P3525_N', 'pCASP9_N', 'PSD95_N', 'SNCA_N',
       'Ubiquitin_N', 'pGSK3B_Tyr216_N', 'SHH_N', 'BAD_N', 'BCL2_N', 'pS6_N',
       'pCFOS_N', 'SYP_N', 'H3AcK18_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N']]
Y_df=df['mice_class']

In [47]:
##checking accuracy 
score_df = acc_score(X_df,Y_df)
score_df

Classifier  Accuracy
0  RandomForest  0.959259
1     LinearSVM  0.948148
2    KNeighbors  0.948148
3      Logistic  0.903704
4  DecisionTree  0.870370
5     RadialSVM  0.714815
6      AdaBoost  0.300000

#### Choosing the best classifier for further calculations

In [48]:
logmodel = RandomForestClassifier(n_estimators=20, random_state=0)
X_train,X_test, Y_train, Y_test = split(X_df,Y_df)
chromo_df,score_df=generations(X_df,Y_df,size=80,n_feat=X_df.shape[1],n_parents=64,mutation_rate=0.20,n_gen=5,
                         X_train = X_train,X_test = X_test,Y_train = Y_train,Y_test = Y_test)

Best score in generation 1 : [0.9962962962962963]
Best score in generation 2 : [0.9962962962962963]
Best score in generation 3 : [1.0]
Best score in generation 4 : [0.9925925925925926]
Best score in generation 5 : [0.9962962962962963]


In [49]:
imp_features_df=[]
for i in range(len(X_df.columns)):
    if(chromo_df[2][i]):
        imp_features_df.append(X_df.columns[i])
print(imp_features_df)
        

['NR1_N', 'NR2A_N', 'pAKT_N', 'pBRAF_N', 'pCAMKII_N', 'pJNK_N', 'pMEK_N', 'pNR2A_N', 'pRSK_N', 'CREB_N', 'ELK_N', 'GSK3B_N', 'JNK_N', 'MEK_N', 'TRKA_N', 'RSK_N', 'Bcatenin_N', 'SOD1_N', 'DSCR1_N', 'pP70S6_N', 'NUMB_N', 'P70S6_N', 'pGSK3B_N', 'pPKCG_N', 'CDK5_N', 'ADARB1_N', 'ERBB4_N', 'nNOS_N', 'Tau_N', 'GFAP_N', 'GluR4_N', 'P3525_N', 'pCASP9_N', 'pGSK3B_Tyr216_N', 'BAD_N', 'BCL2_N', 'SYP_N', 'H3MeK4_N', 'CaNA_N']


In [50]:
len(imp_features_df)

39

In [51]:
from genetic_selection import GeneticSelectionCV
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
estimator = DecisionTreeClassifier()
model = GeneticSelectionCV(
    estimator, cv=5, verbose=0,
    scoring="accuracy", max_features=50,
    n_population=100, crossover_proba=0.5,
    mutation_proba=0.2, n_generations=50,
    crossover_independent_proba=0.5,
    mutation_independent_proba=0.04,
    tournament_size=3, n_gen_no_change=10,
    caching=True, n_jobs=-1)
model = model.fit(X_df, Y_df)
print('Features:', X_df.columns[model.support_])

Features: Index(['ITSN1_N', 'NR1_N', 'NR2A_N', 'pCAMKII_N', 'PKCA_N', 'pPKCAB_N',
       'BRAF_N', 'GSK3B_N', 'JNK_N', 'MEK_N', 'APP_N', 'SOD1_N', 'MTOR_N',
       'P38_N', 'pMTOR_N', 'pNUMB_N', 'P70S6_N', 'pPKCG_N', 'S6_N', 'ADARB1_N',
       'nNOS_N', 'Tau_N', 'GluR4_N', 'SNCA_N', 'EGR1_N', 'CaNA_N'],
      dtype='object')


### Comparing results obtained by manual and sklearn genetic algorithm 

In [52]:
df_manual=df[['DYRK1A_N', 'ITSN1_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pCAMKII_N', 'pCREB_N', 'pJNK_N', 'pNR2A_N', 'pRSK_N', 'AKT_N', 'BRAF_N', 'CREB_N', 'ERK_N', 'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'AMPKA_N', 'NR2B_N', 'pNUMB_N', 'TIAM1_N', 'pP70S6_N', 'pPKCG_N', 'S6_N', 'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N', 'ERBB4_N', 'nNOS_N', 'Tau_N', 'GFAP_N', 'GluR3_N', 'GluR4_N', 'IL1B_N', 'SNCA_N', 'Ubiquitin_N', 'SHH_N', 'pCFOS_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N','mice_class']]
df_manual.head()

DYRK1A_N   ITSN1_N     NR1_N    NR2A_N    pAKT_N  pCAMKII_N   pCREB_N  \
0  0.503644  0.747193  2.816329  5.990152  0.218830   2.373744  0.232224   
1  0.514617  0.689064  2.789514  5.685038  0.211636   2.292150  0.226972   
2  0.509183  0.730247  2.687201  5.622059  0.209011   2.283337  0.230247   
3  0.442107  0.617076  2.466947  4.979503  0.222886   2.152301  0.207004   
4  0.434940  0.617430  2.365785  4.718679  0.213106   2.134014  0.192158   

     pJNK_N   pNR2A_N    pRSK_N     AKT_N    BRAF_N    CREB_N     ERK_N  \
0  0.306382  0.605673  0.441599  0.859366  0.416289  0.178944  3.685247   
1  0.299051  0.587559  0.445222  0.834659  0.400364  0.173680  3.485287   
2  0.291276  0.602449  0.467668  0.814329  0.399847  0.173905  3.571456   
3  0.296729  0.520293  0.477671  0.727705  0.385639  0.179449  2.970137   
4  0.286961  0.507990  0.483416  0.687794  0.367531  0.174836  2.896334   

      MEK_N    TRKA_N     RSK_N     APP_N  Bcatenin_N   AMPKA_N    NR2B_N  \
0  0.319677  0.813866  0.165846  0.453910    3.037621  0.448099  0.586271   
1  0.304419  0.780504  0.157194  0.430940    2.921882  0.420876  0.545097   
2  0.311747  0.785154  0.160895  0.423187    2.944136  0.404630  0.552994   
3  0.279218  0.734492  0.162210  0.410615    2.500204  0.386860  0.547849   
4  0.273667  0.702699  0.154827  0.398550    2.456560  0.360816  0.512824   

    pNUMB_N   TIAM1_N  pP70S6_N   pPKCG_N      S6_N  ADARB1_N  AcetylH3K9_N  \
0  0.394721  0.482864  0.294170  1.443091  0.354605  1.339070      0.170119   
1  0.368255  0.454519  0.276431  1.439460  0.354548  1.306323      0.171427   
2  0.363880  0.447197  0.256648  1.524364  0.386087  1.279600      0.185456   
3  0.366771  0.442650  0.398534  1.612382  0.290680  1.198765      0.159799   
4  0.351551  0.419095  0.393447  1.645807  0.309345  1.206995      0.164650   

     RRP1_N   ERBB4_N    nNOS_N     Tau_N    GFAP_N   GluR3_N   GluR4_N  \
0  0.159102  0.144989  0.176668  0.125190  0.115291  0.228043  0.142756   
1  0.158129  0.150471  0.178309  0.134275  0.118235  0.238073  0.142037   
2  0.148696  0.145330  0.176213  0.132560  0.117760  0.244817  0.142445   
3  0.166112  0.140656  0.163804  0.123210  0.117439  0.234947  0.145068   
4  0.160687  0.141983  0.167710  0.136838  0.116048  0.255528  0.140871   

     IL1B_N    SNCA_N  Ubiquitin_N     SHH_N   pCFOS_N    EGR1_N  H3MeK4_N  \
0  0.430957  0.108234     1.044979  0.188852  0.108336  0.131790  0.128186   
1  0.457156  0.109749     1.009883  0.200404  0.104315  0.135103  0.131119   
2  0.510472  0.108196     0.996848  0.193685  0.106219  0.133362  0.127431   
3  0.430996  0.119883     0.990225  0.192112  0.111262  0.147444  0.146901   
4  0.481227  0.119524     0.997775  0.205604  0.110694  0.140314  0.148380   

     CaNA_N  mice_class  
0  1.675652           0  
1  1.743610           0  
2  1.926427           0  
3  1.700563           0  
4  1.839730           0

In [53]:
df_manual.columns

Index(['DYRK1A_N', 'ITSN1_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pCAMKII_N',
       'pCREB_N', 'pJNK_N', 'pNR2A_N', 'pRSK_N', 'AKT_N', 'BRAF_N', 'CREB_N',
       'ERK_N', 'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'AMPKA_N',
       'NR2B_N', 'pNUMB_N', 'TIAM1_N', 'pP70S6_N', 'pPKCG_N', 'S6_N',
       'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N', 'ERBB4_N', 'nNOS_N', 'Tau_N',
       'GFAP_N', 'GluR3_N', 'GluR4_N', 'IL1B_N', 'SNCA_N', 'Ubiquitin_N',
       'SHH_N', 'pCFOS_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N', 'mice_class'],
      dtype='object')

In [54]:
df_manual_X=df_manual[['DYRK1A_N', 'ITSN1_N', 'NR1_N', 'NR2A_N', 'pAKT_N', 'pCAMKII_N',
       'pCREB_N', 'pJNK_N', 'pNR2A_N', 'pRSK_N', 'AKT_N', 'BRAF_N', 'CREB_N',
       'ERK_N', 'MEK_N', 'TRKA_N', 'RSK_N', 'APP_N', 'Bcatenin_N', 'AMPKA_N',
       'NR2B_N', 'pNUMB_N', 'TIAM1_N', 'pP70S6_N', 'pPKCG_N', 'S6_N',
       'ADARB1_N', 'AcetylH3K9_N', 'RRP1_N', 'ERBB4_N', 'nNOS_N', 'Tau_N',
       'GFAP_N', 'GluR3_N', 'GluR4_N', 'IL1B_N', 'SNCA_N', 'Ubiquitin_N',
       'SHH_N', 'pCFOS_N', 'EGR1_N', 'H3MeK4_N', 'CaNA_N']]
ss = StandardScaler()
df_manual_X= ss.fit_transform(df_manual_X)
df_manual_X

array([[ 0.31271112,  0.5179336 ,  1.49736237, ..., -1.41662394,
        -1.60789061,  1.06590091],
       [ 0.35679793,  0.28650133,  1.42000895, ..., -1.32521803,
        -1.54684392,  1.28029118],
       [ 0.33496588,  0.45046461,  1.12486049, ..., -1.37325709,
        -1.62359464,  1.85703831],
       ...,
       [-0.79192771, -0.88354273, -1.62725109, ...,  1.27078193,
         3.11724261,  0.29352469],
       [-0.82188815, -0.8130138 , -1.21425893, ...,  1.88117889,
         3.32828966,  0.2089962 ],
       [-0.49491588, -0.62125474, -0.58987222, ...,  1.92748438,
         3.32672533,  0.10478825]])

In [55]:
#Splitting the dataset into  training and validation sets
df_manual_Y=df_manual['mice_class']
X_train, X_test, Y_train, Y_test = train_test_split(df_manual_X, df_manual_Y, test_size=0.2,
                                                    random_state=1,stratify=df_manual_Y) #using stratify to preserve the percentage of data in each class

In [56]:
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(n_estimators=150, random_state=0)
random_forest.fit(X_train, Y_train)
Y_pred = random_forest.predict(X_test)
accuracy_score(Y_test,Y_pred)

1.0

100% accuracy achieved 

In [57]:
df_library=df[['DYRK1A_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pCAMKII_N', 'pJNK_N',
       'PKCA_N', 'pNR2B_N', 'pPKCAB_N', 'ERK_N', 'GSK3B_N', 'APP_N', 'SOD1_N',
       'P38_N', 'NR2B_N', 'NUMB_N', 'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N',
       'ERBB4_N', 'nNOS_N', 'GluR3_N', 'IL1B_N', 'pGSK3B_Tyr216_N', 'SHH_N',
       'pCFOS_N', 'H3AcK18_N', 'EGR1_N', 'CaNA_N','mice_class']]
df_library.head()

DYRK1A_N    BDNF_N     NR1_N    NR2A_N  pCAMKII_N    pJNK_N    PKCA_N  \
0  0.503644  0.430175  2.816329  5.990152   2.373744  0.306382  0.402698   
1  0.514617  0.411770  2.789514  5.685038   2.292150  0.299051  0.385987   
2  0.509183  0.418309  2.687201  5.622059   2.283337  0.291276  0.381002   
3  0.442107  0.358626  2.466947  4.979503   2.152301  0.296729  0.377087   
4  0.434940  0.358802  2.365785  4.718679   2.134014  0.286961  0.363502   

    pNR2B_N  pPKCAB_N     ERK_N   GSK3B_N     APP_N    SOD1_N     P38_N  \
0  1.877684  2.308745  3.685247  1.537227  0.453910  0.369510  0.335336   
1  1.725774  2.043037  3.485287  1.509249  0.430940  0.342279  0.324835   
2  1.731873  2.017984  3.571456  1.501244  0.423187  0.343696  0.324852   
3  1.566852  2.132754  2.970137  1.419710  0.410615  0.344509  0.330121   
4  1.480059  2.013697  2.896334  1.359876  0.398550  0.329126  0.313415   

     NR2B_N    NUMB_N   pPKCG_N    CDK5_N      S6_N  ADARB1_N   ERBB4_N  \
0  0.586271  0.182150  1.443091  0.294700  0.354605  1.339070  0.144989   
1  0.545097  0.182086  1.439460  0.294060  0.354548  1.306323  0.150471   
2  0.552994  0.184388  1.524364  0.301881  0.386087  1.279600  0.145330   
3  0.547849  0.161768  1.612382  0.296382  0.290680  1.198765  0.140656   
4  0.512824  0.160200  1.645807  0.296829  0.309345  1.206995  0.141983   

     nNOS_N   GluR3_N    IL1B_N  pGSK3B_Tyr216_N     SHH_N   pCFOS_N  \
0  0.176668  0.228043  0.430957         0.831557  0.188852  0.108336   
1  0.178309  0.238073  0.457156         0.849270  0.200404  0.104315   
2  0.176213  0.244817  0.510472         0.846709  0.193685  0.106219   
3  0.163804  0.234947  0.430996         0.833277  0.192112  0.111262   
4  0.167710  0.255528  0.481227         0.878668  0.205604  0.110694   

   H3AcK18_N    EGR1_N    CaNA_N  mice_class  
0   0.114783  0.131790  1.675652           0  
1   0.111974  0.135103  1.743610           0  
2   0.111883  0.133362  1.926427           0  
3   0.130405  0.147444  1.700563           0  
4   0.118481  0.140314  1.839730           0

In [58]:
df_library.columns


Index(['DYRK1A_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pCAMKII_N', 'pJNK_N',
       'PKCA_N', 'pNR2B_N', 'pPKCAB_N', 'ERK_N', 'GSK3B_N', 'APP_N', 'SOD1_N',
       'P38_N', 'NR2B_N', 'NUMB_N', 'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N',
       'ERBB4_N', 'nNOS_N', 'GluR3_N', 'IL1B_N', 'pGSK3B_Tyr216_N', 'SHH_N',
       'pCFOS_N', 'H3AcK18_N', 'EGR1_N', 'CaNA_N', 'mice_class'],
      dtype='object')

In [59]:
df_library_X=df_library[['DYRK1A_N', 'BDNF_N', 'NR1_N', 'NR2A_N', 'pCAMKII_N', 'pJNK_N',
       'PKCA_N', 'pNR2B_N', 'pPKCAB_N', 'ERK_N', 'GSK3B_N', 'APP_N', 'SOD1_N',
       'P38_N', 'NR2B_N', 'NUMB_N', 'pPKCG_N', 'CDK5_N', 'S6_N', 'ADARB1_N',
       'ERBB4_N', 'nNOS_N', 'GluR3_N', 'IL1B_N', 'pGSK3B_Tyr216_N', 'SHH_N',
       'pCFOS_N', 'H3AcK18_N', 'EGR1_N', 'CaNA_N']]
ss = StandardScaler()
df_library_X= ss.fit_transform(df_library_X)
df_library_X

array([[ 0.31271112,  2.2536689 ,  1.49736237, ..., -1.01161458,
        -1.41662394,  1.06590091],
       [ 0.35679793,  1.8802795 ,  1.42000895, ..., -1.06345775,
        -1.32521803,  1.28029118],
       [ 0.33496588,  2.01292763,  1.12486049, ..., -1.06512993,
        -1.37325709,  1.85703831],
       ...,
       [-0.79192771, -1.72382963, -1.62725109, ...,  2.79900403,
         1.27078193,  0.29352469],
       [-0.82188815, -1.52387571, -1.21425893, ...,  3.06895768,
         1.88117889,  0.2089962 ],
       [-0.49491588, -1.26845332, -0.58987222, ...,  3.05282386,
         1.92748438,  0.10478825]])

In [60]:
#Splitting the dataset into  training and validation sets
df_library_Y=df_library['mice_class']
X_train, X_test, Y_train, Y_test = train_test_split(df_library_X, df_library_Y, test_size=0.2,
                                                    random_state=1,stratify=df_library_Y) #using stratify to preserve the percentage of data in each class

In [61]:
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(n_estimators=150, random_state=0)
random_forest.fit(X_train, Y_train)
Y_pred = random_forest.predict(X_test)
accuracy_score(Y_test,Y_pred)

0.9953703703703703

Slightly less than 100% accuracy achieved

In [68]:
df_manual.head()

DYRK1A_N   ITSN1_N     NR1_N    NR2A_N    pAKT_N  pCAMKII_N   pCREB_N  \
0  0.503644  0.747193  2.816329  5.990152  0.218830   2.373744  0.232224   
1  0.514617  0.689064  2.789514  5.685038  0.211636   2.292150  0.226972   
2  0.509183  0.730247  2.687201  5.622059  0.209011   2.283337  0.230247   
3  0.442107  0.617076  2.466947  4.979503  0.222886   2.152301  0.207004   
4  0.434940  0.617430  2.365785  4.718679  0.213106   2.134014  0.192158   

     pJNK_N   pNR2A_N    pRSK_N     AKT_N    BRAF_N    CREB_N     ERK_N  \
0  0.306382  0.605673  0.441599  0.859366  0.416289  0.178944  3.685247   
1  0.299051  0.587559  0.445222  0.834659  0.400364  0.173680  3.485287   
2  0.291276  0.602449  0.467668  0.814329  0.399847  0.173905  3.571456   
3  0.296729  0.520293  0.477671  0.727705  0.385639  0.179449  2.970137   
4  0.286961  0.507990  0.483416  0.687794  0.367531  0.174836  2.896334   

      MEK_N    TRKA_N     RSK_N     APP_N  Bcatenin_N   AMPKA_N    NR2B_N  \
0  0.319677  0.813866  0.165846  0.453910    3.037621  0.448099  0.586271   
1  0.304419  0.780504  0.157194  0.430940    2.921882  0.420876  0.545097   
2  0.311747  0.785154  0.160895  0.423187    2.944136  0.404630  0.552994   
3  0.279218  0.734492  0.162210  0.410615    2.500204  0.386860  0.547849   
4  0.273667  0.702699  0.154827  0.398550    2.456560  0.360816  0.512824   

    pNUMB_N   TIAM1_N  pP70S6_N   pPKCG_N      S6_N  ADARB1_N  AcetylH3K9_N  \
0  0.394721  0.482864  0.294170  1.443091  0.354605  1.339070      0.170119   
1  0.368255  0.454519  0.276431  1.439460  0.354548  1.306323      0.171427   
2  0.363880  0.447197  0.256648  1.524364  0.386087  1.279600      0.185456   
3  0.366771  0.442650  0.398534  1.612382  0.290680  1.198765      0.159799   
4  0.351551  0.419095  0.393447  1.645807  0.309345  1.206995      0.164650   

     RRP1_N   ERBB4_N    nNOS_N     Tau_N    GFAP_N   GluR3_N   GluR4_N  \
0  0.159102  0.144989  0.176668  0.125190  0.115291  0.228043  0.142756   
1  0.158129  0.150471  0.178309  0.134275  0.118235  0.238073  0.142037   
2  0.148696  0.145330  0.176213  0.132560  0.117760  0.244817  0.142445   
3  0.166112  0.140656  0.163804  0.123210  0.117439  0.234947  0.145068   
4  0.160687  0.141983  0.167710  0.136838  0.116048  0.255528  0.140871   

     IL1B_N    SNCA_N  Ubiquitin_N     SHH_N   pCFOS_N    EGR1_N  H3MeK4_N  \
0  0.430957  0.108234     1.044979  0.188852  0.108336  0.131790  0.128186   
1  0.457156  0.109749     1.009883  0.200404  0.104315  0.135103  0.131119   
2  0.510472  0.108196     0.996848  0.193685  0.106219  0.133362  0.127431   
3  0.430996  0.119883     0.990225  0.192112  0.111262  0.147444  0.146901   
4  0.481227  0.119524     0.997775  0.205604  0.110694  0.140314  0.148380   

     CaNA_N  mice_class  
0  1.675652           0  
1  1.743610           0  
2  1.926427           0  
3  1.700563           0  
4  1.839730           0

**We see on our dataset with all previous processing performed and using the selected features by genetic algorithm we are achieving 100% accuracy, now lets try to implement XGboost to compare its performance with random forest ensemble algorithm.**

In [73]:
#XGBOOST WITHOUT PARAMETER TUNING
import xgboost as xgb

from sklearn import datasets

from sklearn.model_selection import train_test_split

from sklearn.preprocessing import LabelEncoder 

from xgboost import XGBClassifier

from sklearn.metrics import accuracy_score

from sklearn.model_selection import GridSearchCV

In [66]:
#Splitting the dataset into  training and validation sets
df_manual_Y=df_manual['mice_class']
X_train, X_test, Y_train, Y_test = train_test_split(df_manual_X, df_manual_Y, test_size=0.2,
                                                    random_state=1,stratify=df_manual_Y) #using stratify to preserve the percentage of data in each class

In [70]:
model = XGBClassifier() 
model.fit(X_train, Y_train)

XGBClassifier(base_score=0.5, booster='gbtree', callbacks=None,
              colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
              early_stopping_rounds=None, enable_categorical=False,
              eval_metric=None, gamma=0, gpu_id=-1, grow_policy='depthwise',
              importance_type=None, interaction_constraints='',
              learning_rate=0.300000012, max_bin=256, max_cat_to_onehot=4,
              max_delta_step=0, max_depth=6, max_leaves=0, min_child_weight=1,
              missing=nan, monotone_constraints='()', n_estimators=100,
              n_jobs=0, num_parallel_tree=1, objective='multi:softprob',
              predictor='auto', random_state=0, reg_alpha=0, ...)

In [72]:
y_pred = model.predict(X_test) 
predictions = [round(value) for value in y_pred]
accuracy = accuracy_score(Y_test, predictions) 
print("Accuracy: %.2f%%" % (accuracy * 100.0))

Accuracy: 96.30%


**Without parameter tuning it is possible to achieve 96.3% accuracy using the XGboost gradient boosting algorithm.**

In [96]:
#orignally we are obtaining 96.30% accuracy
#performing parameter tuning
#Grid Search for XGBoost
# Define the search space
# Hyperparameter tuning
from sklearn.model_selection import StratifiedKFold, GridSearchCV

param_grid = { 
    # Percentage of columns to be randomly samples for each tree.
    "colsample_bytree": [ 0.3, 0.5 , 0.8 ],
    # reg_alpha provides l1 regularization to the weight, higher values result in more conservative models
    "reg_alpha": [0, 0.5, 1, 5],
    # reg_lambda provides l2 regularization to the weight, higher values result in more conservative models
    "reg_lambda": [0, 0.5, 1, 5]
    }
# Set up score
scoring = ['accuracy']
# Set up the k-fold cross-validation
kfold = StratifiedKFold(n_splits=3, shuffle=True, random_state=0)
# Define grid search
grid_search = GridSearchCV(estimator=model, 
                           param_grid=param_grid, 
                           scoring=scoring, 
                           refit='accuracy', 
                           n_jobs=-1, 
                           cv=kfold, 
                           verbose=0)
# Fit grid search
grid_result = grid_search.fit(X_train, Y_train)
# Print grid search summary
grid_result
# Print the best score and the corresponding hyperparameters
#print(f'The best score is {grid_result.best_score_:.4f}')
print(f'The best hyperparameters are {grid_result.best_params_}')
# Make prediction using the best model
grid_predict = grid_search.predict(X_test)
# Get predicted probabilities
grid_predict_prob = grid_search.predict_proba(X_test)[:,1]
# Get performance metrics
print("Accuracy post hyperparameter tuning: ",accuracy_score(Y_test, grid_predict)*100,"%")

The best hyperparameters are {'colsample_bytree': 0.3, 'reg_alpha': 0, 'reg_lambda': 0}
Accuracy post hyperparameter tuning:  98.61111111111111 %


**On performing hyperparameter tuning using GridSearch we are able to achieve more than 2% improvement in accuracy from 96.3%  to 98.61%. Although this is lower than the 100% accuracy obtained by using random forests.**